src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/Registry/CosmosDbSearchParameterStatusDataStore.cs

```diff
@@ -34,43 +36,82 @@ public class CosmosDbSearchParameterStatusDataStore : ISearchParameterStatusData
 
-                    parameterStatus.AddRange(results.Select(x => x.ToSearchParameterStatus()));
+                    if (lastUpdated < _lastRefreshed)
+                    {
+                        return _statusList;
+                    }
                 }
-                while (query.HasMoreResults);
 
-                if (!parameterStatus.Any())
+                var parameterStatus = new List<ResourceSearchParameterStatus>();
+
+                do
                 {
-                    await Task.Delay(TimeSpan.FromSeconds(1), cancellationSource.Token);
+                    var query = _queryFactory.Create<SearchParameterStatusWrapper>(
+                        clientScope.Value,
+                        new CosmosQueryContext(
+                            new QueryDefinition("select * from c"),
+                            new QueryRequestOptions
+                            {
+                                PartitionKey = new PartitionKey(SearchParameterStatusWrapper.SearchParameterStatusPartitionKey),
+                            }));
+
+                    do
+                    {
+                        FeedResponse<SearchParameterStatusWrapper> results = await query.ExecuteNextAsync();
+
+                        parameterStatus.AddRange(results.Select(x => x.ToSearchParameterStatus()));
+                    }
+                    while (query.HasMoreResults);
+
+                    if (!parameterStatus.Any())
+                    {
+                        await Task.Delay(TimeSpan.FromSeconds(1), cancellationSource.Token);
+                    }
                 }
+                while (!parameterStatus.Any() && !cancellationSource.IsCancellationRequested);
+
+                _lastRefreshed = startedCheck;
+                _statusList = parameterStatus;
+
+                return _statusList;
             }
-            while (!parameterStatus.Any() && !cancellationSource.IsCancellationRequested);
+            finally
+            {
+                _statusListSemaphore.Release();
+            }
+        }
+
+        private async Task<DateTimeOffset> GetSearchParameterStatusesLastUpdatedAsync(IScoped<Container> container, CancellationToken cancellationToken)
+        {
+            var query = _queryFactory.Create<DateTimeOffset>(
+                container.Value,
+                new CosmosQueryContext(
+                    new QueryDefinition("select value c.lastUpdated from c order by c.lastUpdated desc"),
+                    new QueryRequestOptions
+                    {
+                        PartitionKey = new PartitionKey(SearchParameterStatusWrapper.SearchParameterStatusPartitionKey),
+                        MaxItemCount = 1,
+                    }));
 
-            return parameterStatus;
+            var response = await query.ExecuteNextAsync(cancellationToken);
+
+            return response.FirstOrDefault();
```

Comment:
What happens if the query returns no results?

Reply:
It might be DateTimeOffset.MinValue, which should be fine, its only matters when there is a date that is more current